ledger/store/trackerdb/dualdriver/accounts_writer.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package dualdriver
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/ledger/store/trackerdb"
+)
+
+type accountsWriter struct {
+	primary   trackerdb.AccountsWriter
+	secondary trackerdb.AccountsWriter
+}
+
+// Close implements trackerdb.AccountsWriter
+func (aw *accountsWriter) Close() {
+	aw.primary.Close()
+	aw.secondary.Close()
+}
+
+// DeleteAccount implements trackerdb.AccountsWriter
+func (aw *accountsWriter) DeleteAccount(accRef trackerdb.AccountRef) (rowsAffected int64, err error) {
+	// parse ref
+	xRef := accRef.(accountRef)
+	// Note: rowsAffected is ignored because it is not possible to determine this correctly in all engines
+	rowsAffectedP, errP := aw.primary.DeleteAccount(xRef.primary)
+	_, errS := aw.secondary.DeleteAccount(xRef.secondary)
```

Comment:
Not an issue on read commands, but for write commands, when there's any error, it is no longer clear if the update was applied to one, both, or neither database.

That's all fine, but this would also indicate that the database the dual driver controls is now in an unusable state for any further testing, correct? Like, if you failed to delete, any subsequent add/delete you do may continue to diverge. Even if some later commands *are* consistent and do work, the total database is diverged so the behavior also diverges.

With that in mind, I wonder if the dual driver shouldn't maintain a flag to indicate its internal consistency. If there's a failure on write, the inconsistent flag should go up, and any further calls (read or write) should return via guard clause,

```
if dualdriver.inconsistent {
return ErrInconsistentDatabase
}
```

The situation I'd be concerned about is a test which uses the database even after a failure to write, causing harder to understand errors. Returning an error indicating the database isn't usable would keep developers from blaming the later command.

Reply:
if you are writing tests and ignoring the error returned on the db calls you can run into this case where things worked on one but not the other. 
 
Not sure why you would do such thing. 
In any case, if you are ignoring errors returned, returning yet another error with a different name will likely get ignored too. 

The alternative design here was to just panic and blow up the test on inconsistencies.

Reply:
> The alternative design here was to just panic and blow up the test on inconsistencies.

Right, and that's the only course of action no matter what. It's fine that we expect a test writer to notice their own errors and act accordingly, but if in the engine you have knowledge that your internal state is unusable, then the DB itself knows it doesn't even need to try anymore.